app/src/main/java/com/zeapo/pwdstore/git/sshj/OpenKeychainKeyProvider.kt

```diff
@@ -0,0 +1,213 @@
+
+    private suspend fun prepareAndUse(block: (provider: OpenKeychainKeyProvider) -> Unit) {
+        prepare()
+        use(block)
+    }
+
+    private suspend fun prepare() {
+        continueAfterUserInteraction = withContext(Dispatchers.Main) {
+            activity.registerForActivityResult(ActivityResultContracts.StartIntentSenderForResult()) { result ->
+                currentCont?.let { cont ->
+                    currentCont = null
+                    val data = result.data
+                    if (data != null)
+                        cont.resume(data)
+                    else
+                        cont.resumeWithException(UserAuthException(DisconnectReason.AUTH_CANCELLED_BY_USER))
+                }
+            }
+        }
+
+        sshServiceApi = suspendCoroutine { cont ->
+            sshServiceConnection.connect(object : SshAuthenticationConnection.OnBound {
+                override fun onBound(sshAgent: ISshAuthenticationService) {
+                    d { "Bound to SshAuthenticationApi: $OPENPGP_PROVIDER" }
+                    cont.resume(SshAuthenticationApi(context, sshAgent))
+                }
+
+                override fun onError() {
+                    throw UserAuthException(DisconnectReason.UNKNOWN, "OpenKeychain service unavailable")
+                }
+            })
+        }
+
+        if (keyId == null) {
+            selectKey()
+        }
+        check(keyId != null)
+        fetchPublicKey()
+        makePrivateKey()
+    }
+
+    private suspend fun fetchPublicKey(isRetry: Boolean = false) {
+        when (val sshPublicKeyResponse = executeApiRequest(SshPublicKeyRequest(keyId))) {
+            is ApiResponse.Success -> {
+                val response = sshPublicKeyResponse.response as SshPublicKeyResponse
+                val sshPublicKey = response.sshPublicKey!!
+                val sshKeyParts = sshPublicKey.split("""\s+""".toRegex())
+                check(sshKeyParts.size >= 2) { "OpenKeychain API returned invalid SSH key" }
+                @Suppress("BlockingMethodInNonBlockingContext")
+                publicKey = Buffer.PlainBuffer(Base64.decode(sshKeyParts[1])).readPublicKey()
+            }
+            is ApiResponse.NoSuchKey -> if (isRetry) {
+                throw sshPublicKeyResponse.exception
+            } else {
+                // Allow the user to reselect an authentication key and retry
+                selectKey()
+                fetchPublicKey(true)
+            }
+            is ApiResponse.GeneralError -> throw sshPublicKeyResponse.exception
+        }
+    }
+
+    private suspend fun selectKey() {
+        when (val keySelectionResponse = executeApiRequest(KeySelectionRequest())) {
+            is ApiResponse.Success -> keyId = (keySelectionResponse.response as KeySelectionResponse).keyId
+            is ApiResponse.GeneralError -> throw keySelectionResponse.exception
+            is ApiResponse.NoSuchKey -> throw keySelectionResponse.exception
```

Comment:
would it make sense to collapse these into a single when branch?

Reply:
That doesn't work since the smart cast algorithm isn't smart enough to handle different types that have a member of the same name and type.